src/systems/physics/Physics.cc

```diff
@@ -1280,6 +1321,29 @@ void PhysicsPrivate::CreateCollisionEntities(const EntityComponentManager &_ecm,
         }
         auto linkPtrPhys = this->entityLinkMap.Get(_parent->Data());
 
+        if (const auto existingShape = linkPtrPhys->GetShape(_name->Data()))
+        {
+          // No need to create this collision shape because it was already
+          // created when parsing the model.
+
+          // TODO(MXG): There's some significant design inconsistency in this
+          // CreateCollisionEntities function. There's a mix of using the
+          // ConstructSdfCollision feature and other collision attachment
+          // features like AttachHeightMapFeature and AttachMeshShape. This
+          // makes it confusing for physics plugin implementers to know where
+          // and how to implement the construction of collision shapes.
+          //
+          // In the current implementation, if a physics engine implements
+          // constructing a mesh or heightmap in ConstructSdfCollision then that
+          // ability won't actually get leveraged by gz-sim because it
+          // specifically uses AttachMeshShapeFeature and AttachHeightMapFeature
+          // for those shapes. The physics plugin implementer needs to know to
+          // implement those specific features instead.
```

Comment:
the reason for this is that the mesh and heightmap SDFormat elements contain URI's that in many cases point to fuel resources, but libsdformat doesn't depend on fuel. The `Attach*ShapeFeature`s accept `gz::common` types that contain the mesh and heightmap info, and `gz-sim` is responsible for fetching the data and populating those types.

Properly implementing the `ConstructSdfMesh` and `ConstructSdfHeightmap` would involve resolving those URIs, which seems beyond the responsibility of a physics plugin, so perhaps we should just deprecate those features from gz-physics

Reply:
Yeah, I think the motivation totally makes sense.

I think what I'd propose is to move the implementations of the uri resolving functionality into a reusable utility function in the `SDFormat` library. Maybe even have them as methods in `MeshShape` and `HeightmapShape` classes for maximum visibility. The utility functions could just output the same data structures that are being passed into `AttachMeshShape` and `AttachHeightmapShape` so the physics plugin developer can easily recycle the implementations of those functions for `ConstructSdfCollision`.

Reply:
I've been confused by this in the past (https://github.com/gazebosim/gz-physics/issues/9):

"_For example, what does [ConstructCollision](https://github.com/ignitionrobotics/ign-physics/blob/a99c99d267972afee4501c061e31dd9821fb131d/sdf/include/ignition/physics/sdf/ConstructCollision.hh#L36) do? Does it create and attach a collision to a link? How is that different from [AttachMeshShape](https://github.com/ignitionrobotics/ign-physics/blob/a99c99d267972afee4501c061e31dd9821fb131d/mesh/include/ignition/physics/mesh/MeshShape.hh#L106)? If they do similar things, why are their names so different?_"

> mesh and heightmap SDFormat elements contain URI's that in many cases point to fuel resources, but libsdformat doesn't depend on fuel

SDF has the ability of resolving Fuel URIs for `<include>`s, we pass it a Fuel-Tools callback so it knows how to resolve those:

https://github.com/gazebosim/gz-sim/blob/417609c88d7110864a7a10ae7ea9f6d7ebf924ab/src/Server.cc#L66-L68

So it could also resolve the URIs for other assets.

I think the main problem is that SDFormat doesn't depend on Gazebo Common, which is a huge dependency and brings external graphics libraries, so it can't return  `gz::common` types.

I guess one option is passing those resolved URIs to `gz-rendering` and `gz-physics`, and each implementation decides how to load the assets. They can use `gz::common` if they want. It may cause a lot of duplication across various plugins though, or inconsistencies on how the meshes are loaded if they use different loaders.

Reply:
> I think the main problem is that SDFormat doesn't depend on Gazebo Common, which is a huge dependency and brings external graphics libraries, so it can't return `gz::common` types.

yeah, I think this is the real crux of the problem. Perhaps we could add basic Mesh and HeightmapData classes to gz-math, while keeping the parsers in gz-common? Then libsdformat could return the gz-math types, and the find-file callback function could use fuel and gz-common's parsers to resolve the URI into the math object

Reply:
I think @mjcarroll has a solution with https://github.com/gazebosim/sdformat/pull/1147 so that the SDF DOM object passed into gz-physics has all the resource paths resolved.

Reply:
> I think @mjcarroll has a solution with [gazebosim/sdformat#1147](https://github.com/gazebosim/sdformat/pull/1147) so that the SDF DOM object passed into gz-physics has all the resource paths resolved.

that means we don't need to use fuel, but it's still just a file path provided by `libsdformat`, but something will still need to parse that file, and in general we can't assume that the physics implementation will be able to parse it, so I think that wouldn't resolve the matter of properly implementing `ConstructSdfMesh` or `ConstructSdfHeightmap` in the general case so that `AttachMeshShape` and `AttacheHeightmapShape` aren't needed